File: backend/schemas/ab_testing.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional
from datetime import datetime
from enum import Enum

class TestType(str, Enum):
    RAG_VS_BASELINE = "rag_vs_baseline"
    MODEL_COMPARISON = "model_comparison"
    RETRIEVAL_METHOD = "retrieval_method"
    CHUNK_SIZE = "chunk_size"
    CONTEXT_WINDOW = "context_window"

class TestStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    ANALYZED = "analyzed"

# AB测试基础模式
class ABTestBase(BaseModel):
    name: str = Field(..., description="测试名称")
    description: Optional[str] = Field(None, description="测试描述")
    test_type: TestType = Field(..., description="测试类型")
    traffic_split: float = Field(0.5, ge=0.1, le=0.9, description="流量分配比例")
    sample_size: Optional[int] = Field(None, description="样本大小")
    duration_days: Optional[int] = Field(None, description="测试持续天数")

class ABTestCreate(ABTestBase):
    group_a_config: Dict[str, Any] = Field(..., description="A组配置")
    group_b_config: Dict[str, Any] = Field(..., description="B组配置")
    config: Optional[Dict[str, Any]] = Field(None, description="通用配置")

class ABTestUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[TestStatus] = None
    group_a_config: Optional[Dict[str, Any]] = None
    group_b_config: Optional[Dict[str, Any]] = None
    config: Optional[Dict[str, Any]] = None

class ABTestResponse(ABTestBase):
    id: int
    status: TestStatus
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    group_a_config: Dict[str, Any]
    group_b_config: Dict[str, Any]
    config: Optional[Dict[str, Any]] = None

    class Config:
        from_attributes = True

# 测试会话模式
class ABTestSessionBase(BaseModel):
    test_id: int
    user_id: int
    session_id: str
    group: str
    group_config: Dict[str, Any]

class ABTestSessionCreate(ABTestSessionBase):
    pass

class ABTestSessionResponse(ABTestSessionBase):
    id: int
    start_time: datetime
    end_time: Optional[datetime] = None
    duration_seconds: Optional[int] = None
    total_queries: int
    successful_queries: int
    failed_queries: int

    class Config:
        from_attributes = True

# 测试交互模式
class ABTestInteractionBase(BaseModel):
    session_id: int
    query: str
    response: Optional[str] = None
    response_time_ms: Optional[int] = None
    retrieved_chunks: Optional[List[Dict[str, Any]]] = None
    retrieval_time_ms: Optional[int] = None
    chunk_count: Optional[int] = None
    model_used: Optional[str] = None
    tokens_used: Optional[int] = None
    relevance_score: Optional[float] = None
    accuracy_score: Optional[float] = None
    helpfulness_score: Optional[float] = None
    user_rating: Optional[int] = Field(None, ge=1, le=5)
    user_feedback: Optional[str] = None

class ABTestInteractionCreate(ABTestInteractionBase):
    pass

class ABTestInteractionResponse(ABTestInteractionBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True

# 测试指标模式
class ABTestMetricBase(BaseModel):
    test_id: int
    metric_name: str
    metric_type: str
    group_a_value: Optional[float] = None
    group_b_value: Optional[float] = None
    difference: Optional[float] = None
    improvement_percentage: Optional[float] = None
    confidence_level: Optional[float] = None
    p_value: Optional[float] = None
    is_significant: Optional[bool] = None

class ABTestMetricResponse(ABTestMetricBase):
    id: int
    calculated_at: datetime

    class Config:
        from_attributes = True

# 测试问题模式
class TestQuestionBase(BaseModel):
    test_id: int
    question: str
    category: Optional[str] = None
    difficulty: Optional[str] = None
    expected_answer: Optional[str] = None
    key_points: Optional[List[str]] = None
    scoring_criteria: Optional[Dict[str, Any]] = None

class TestQuestionCreate(TestQuestionBase):
    pass

class TestQuestionResponse(TestQuestionBase):
    id: int

    class Config:
        from_attributes = True

# 问题评估模式
class QuestionEvaluationBase(BaseModel):
    question_id: int
    interaction_id: int
    evaluator_id: int
    relevance_score: Optional[float] = Field(None, ge=0, le=10)
    accuracy_score: Optional[float] = Field(None, ge=0, le=10)
    completeness_score: Optional[float] = Field(None, ge=0, le=10)
    clarity_score: Optional[float] = Field(None, ge=0, le=10)
    helpfulness_score: Optional[float] = Field(None, ge=0, le=10)
    overall_score: Optional[float] = Field(None, ge=0, le=10)
    strengths: Optional[str] = None
    weaknesses: Optional[str] = None
    suggestions: Optional[str] = None

class QuestionEvaluationCreate(QuestionEvaluationBase):
    pass

class QuestionEvaluationResponse(QuestionEvaluationBase):
    id: int
    evaluated_at: datetime

    class Config:
        from_attributes = True

# 测试摘要模式
class ABTestSummary(BaseModel):
    test_id: int
    test_name: str
    status: TestStatus
    total_sessions: int
    group_a_sessions: int
    group_b_sessions: int
    total_interactions: int
    successful_interactions: int
    success_rate: float
    avg_response_time_ms: float
    avg_user_rating: float
    created_at: datetime
    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None

# 测试查询模式
class TestQueryRequest(BaseModel):
    query: str
    kb_id: Optional[int] = None
    session_id: Optional[str] = None

class TestQueryResponse(BaseModel):
    answer: str
    sources: List[Dict[str, Any]]
    response_time_ms: int
    model_used: str
    tokens_used: int
    interaction_id: int

# 用户反馈模式
class UserFeedbackRequest(BaseModel):
    interaction_id: int
    rating: int = Field(..., ge=1, le=5, description="用户评分")
    feedback: Optional[str] = None

# 批量操作模式
class BatchTestCreate(BaseModel):
    tests: List[ABTestCreate]

class BatchQuestionCreate(BaseModel):
    test_id: int
    questions: List[TestQuestionCreate]

# 统计分析模式
class TestStatistics(BaseModel):
    test_id: int
    total_sessions: int
    total_interactions: int
    avg_session_duration: float
    avg_interactions_per_session: float
    success_rate: float
    avg_response_time: float
    avg_user_rating: float
    group_a_stats: Dict[str, Any]
    group_b_stats: Dict[str, Any]
    significant_metrics: List[ABTestMetricResponse]

# 测试配置模板
class TestConfigTemplate(BaseModel):
    name: str
    description: str
    test_type: TestType
    group_a_config: Dict[str, Any]
    group_b_config: Dict[str, Any]
    config: Optional[Dict[str, Any]] = None 

File: frontend/src/api/ab-testing.ts
import api from './index'

// AB测试相关类型定义
export interface ABTest {
  id: number
  name: string
  description?: string
  test_type: 'rag_vs_baseline' | 'model_comparison' | 'retrieval_method' | 'chunk_size' | 'context_window'
  status: 'draft' | 'active' | 'paused' | 'completed' | 'analyzed'
  traffic_split: number
  sample_size?: number
  duration_days?: number
  group_a_config: Record<string, any>
  group_b_config: Record<string, any>
  config?: Record<string, any>
  created_by: number
  created_at: string
  updated_at?: string
  started_at?: string
  ended_at?: string
}

export interface ABTestCreate {
  name: string
  description?: string
  test_type: 'rag_vs_baseline' | 'model_comparison' | 'retrieval_method' | 'chunk_size' | 'context_window'
  traffic_split: number
  sample_size?: number
  duration_days?: number
  group_a_config: Record<string, any>
  group_b_config: Record<string, any>
  config?: Record<string, any>
}

export interface ABTestUpdate {
  name?: string
  description?: string
  status?: 'draft' | 'active' | 'paused' | 'completed' | 'analyzed'
  group_a_config?: Record<string, any>
  group_b_config?: Record<string, any>
  config?: Record<string, any>
}

export interface ABTestSummary {
  test_id: number
  test_name: string
  status: string
  total_sessions: number
  group_a_sessions: number
  group_b_sessions: number
  total_interactions: number
  successful_interactions: number
  success_rate: number
  avg_response_time_ms: number
  avg_user_rating: number
  created_at: string
  started_at?: string
  ended_at?: string
}

export interface ABTestSession {
  id: number
  test_id: number
  user_id: number
  session_id: string
  group: string
  group_config: Record<string, any>
  start_time: string
  end_time?: string
  duration_seconds?: number
  total_queries: number
  successful_queries: number
  failed_queries: number
}

export interface ABTestInteraction {
  id: number
  session_id: number
  query: string
  response?: string
  response_time_ms?: number
  retrieved_chunks?: any[]
  retrieval_time_ms?: number
  chunk_count?: number
  model_used?: string
  tokens_used?: number
  relevance_score?: number
  accuracy_score?: number
  helpfulness_score?: number
  user_rating?: number
  user_feedback?: string
  created_at: string
}

export interface ABTestMetric {
  id: number
  test_id: number
  metric_name: string
  metric_type: string
  group_a_value?: number
  group_b_value?: number
  difference?: number
  improvement_percentage?: number
  confidence_level?: number
  p_value?: number
  is_significant?: boolean
  calculated_at: string
}

export interface TestQuestion {
  id: number
  test_id: number
  question: string
  category?: string
  difficulty?: string
  expected_answer?: string
  key_points?: string[]
  scoring_criteria?: Record<string, any>
}

export interface TestQuestionCreate {
  test_id: number
  question: string
  category?: string
  difficulty?: string
  expected_answer?: string
  key_points?: string[]
  scoring_criteria?: Record<string, any>
}

export interface TestQueryRequest {
  query: string
  kb_id?: number
  session_id?: string
}

export interface TestQueryResponse {
  answer: string
  sources: any[]
  response_time_ms: number
  model_used: string
  tokens_used: number
  interaction_id: number
}

export interface UserFeedbackRequest {
  interaction_id: number
  rating: number
  feedback?: string
}

export interface TestConfigTemplate {
  name: string
  description: string
  test_type: string
  group_a_config: Record<string, any>
  group_b_config: Record<string, any>
  config?: Record<string, any>
}

// AB测试管理API
export const abTestingApi = {
  // 创建AB测试
  createTest: (data: ABTestCreate): Promise<ABTest> => {
    return api.post('/ab-testing/tests', data)
  },

  // 获取AB测试列表
  getTests: (params?: {
    skip?: number
    limit?: number
    status_filter?: string
  }): Promise<ABTest[]> => {
    return api.get('/ab-testing/tests', { params })
  },

  // 获取AB测试详情
  getTest: (testId: number): Promise<ABTest> => {
    return api.get(`/ab-testing/tests/${testId}`)
  },

  // 更新AB测试
  updateTest: (testId: number, data: ABTestUpdate): Promise<ABTest> => {
    return api.put(`/ab-testing/tests/${testId}`, data)
  },

  // 删除AB测试
  deleteTest: (testId: number): Promise<void> => {
    return api.delete(`/ab-testing/tests/${testId}`)
  },

  // 启动AB测试
  startTest: (testId: number): Promise<ABTest> => {
    return api.post(`/ab-testing/tests/${testId}/start`)
  },

  // 暂停AB测试
  pauseTest: (testId: number): Promise<ABTest> => {
    return api.post(`/ab-testing/tests/${testId}/pause`)
  },

  // 完成AB测试
  completeTest: (testId: number): Promise<ABTest> => {
    return api.post(`/ab-testing/tests/${testId}/complete`)
  },

  // 获取测试摘要
  getTestSummary: (testId: number): Promise<ABTestSummary> => {
    return api.get(`/ab-testing/tests/${testId}/summary`)
  },

  // 计算测试指标
  calculateMetrics: (testId: number): Promise<ABTestMetric[]> => {
    return api.post(`/ab-testing/tests/${testId}/calculate-metrics`)
  },

  // 获取测试指标
  getTestMetrics: (testId: number): Promise<ABTestMetric[]> => {
    return api.get(`/ab-testing/tests/${testId}/metrics`)
  },

  // 获取配置模板
  getConfigTemplates: (): Promise<TestConfigTemplate[]> => {
    return api.get('/ab-testing/config-templates')
  }
}

// 测试会话API
export const testSessionApi = {
  // 开始测试会话
  startSession: (testId: number): Promise<ABTestSession> => {
    return api.post(`/ab-testing/tests/${testId}/sessions`)
  },

  // 处理测试查询
  processQuery: (testId: number, data: TestQueryRequest): Promise<TestQueryResponse> => {
    return api.post(`/ab-testing/tests/${testId}/query`, data)
  },

  // 提交用户反馈
  submitFeedback: (interactionId: number, data: UserFeedbackRequest): Promise<void> => {
    return api.post(`/ab-testing/interactions/${interactionId}/feedback`, data)
  }
}

// 测试问题API
export const testQuestionApi = {
  // 创建测试问题
  createQuestions: (testId: number, questions: TestQuestionCreate[]): Promise<TestQuestion[]> => {
    return api.post(`/ab-testing/tests/${testId}/questions`, questions)
  },

  // 获取测试问题列表
  getQuestions: (testId: number): Promise<TestQuestion[]> => {
    return api.get(`/ab-testing/tests/${testId}/questions`)
  }
}

// 测试评估API
export const testEvaluationApi = {
  // 评估回答质量
  evaluateResponse: (data: {
    question_id: number
    interaction_id: number
    evaluator_id: number
    relevance_score?: number
    accuracy_score?: number
    completeness_score?: number
    clarity_score?: number
    helpfulness_score?: number
    overall_score?: number
    strengths?: string
    weaknesses?: string
    suggestions?: string
  }): Promise<any> => {
    return api.post('/ab-testing/evaluations', data)
  }
} 